core/lsp/TypecheckEpochManager.cc

```diff
@@ -0,0 +1,116 @@
+    const u4 invalidator = lspEpochInvalidator.load();
+    const bool slowPathRunning = processing != committed;
+    const bool slowPathIsCanceled = processing != invalidator;
+    return TypecheckingStatus{slowPathRunning, slowPathIsCanceled, committed, processing};
+}
+
+TypecheckEpochManager::TypecheckingStatus TypecheckEpochManager::getStatus() const {
+    absl::MutexLock lock(&epochMutex);
+    return getStatusInternal();
+}
+
+bool TypecheckEpochManager::tryCancelSlowPath(u4 newEpoch) {
+    assertConsistentThread(preprocessThreadId, "TypecheckEpochManager::tryCancelSlowPath", "preprocessThread");
+    absl::MutexLock lock(&epochMutex);
+    const u4 processing = currentlyProcessingLSPEpoch.load();
+    ENFORCE(newEpoch != processing); // This would prevent a cancelation from happening.
+    const u4 committed = lastCommittedLSPEpoch.load();
+    // The second condition should never happen, but guard against it in production.
+    if (processing == committed || newEpoch == processing) {
+        return false;
+    }
+    // Cancel slow path by bumping invalidator.
+    lspEpochInvalidator.store(newEpoch);
+    return true;
+}
+
+bool TypecheckEpochManager::tryCommitEpoch(u4 epoch, bool isCancelable,
+                                           optional<shared_ptr<PreemptionTaskManager>> preemptionManager,
+                                           function<void()> typecheck) {
+    assertConsistentThread(typecheckingThreadId, "TypecheckEpochManager::tryCommitEpoch", "typechecking");
+    if (!isCancelable) {
+        typecheck();
+        return true;
+    }
+
+    // Should have called "startCommitEpoch" *before* this method.
+    ENFORCE(ABSL_TS_UNCHECKED_READ(currentlyProcessingLSPEpoch).load() == epoch);
+    // Typechecking does not run under the mutex, as it would prevent another thread from running `tryCancelSlowPath`
+    // during typechecking.
+    typecheck();
+
+    bool committed = false;
+    {
+        absl::MutexLock lock(&epochMutex);
+        // Try to commit.
+        const u4 processing = currentlyProcessingLSPEpoch.load();
+        const u4 invalidator = lspEpochInvalidator.load();
+        if (processing == invalidator) {
+            ENFORCE(lastCommittedLSPEpoch.load() != processing, "Trying to commit an already-committed epoch.");
+            // OK to commit!
+            lastCommittedLSPEpoch.store(processing);
+            committed = true;
+        } else {
+            // Typechecking was canceled.
+            const u4 lastCommitted = lastCommittedLSPEpoch.load();
+            currentlyProcessingLSPEpoch.store(lastCommitted);
+            lspEpochInvalidator.store(lastCommitted);
+        }
+    }
+
+    if (preemptionManager.has_value()) {
+        // Now that we are no longer running a slow path, run a preemption task that might have snuck in while we were
+        // finishing up. No others can be scheduled.
+        (*preemptionManager)->tryRunScheduledPreemptionTask();
```

Comment:
I'm not 100% happy that `TypecheckEpochManager` knows about `PreemptionTaskManager`, but I don't have a better solution at the moment.